lib/Utils/PlatformWin.cpp

```diff
@@ -735,49 +739,64 @@ static void MergeRanges(ImageRanges &Ranges) {
 static uintptr_t FindEHFrame(uintptr_t Caller) {
   ImageBaseMap& Unwind = getImageBaseMap();
   for (auto&& Itr : Unwind) {
-    const uintptr_t CurAddr = Itr.first;
+    const uintptr_t ImgBase = Itr.first;
     for (auto&& Rng : Itr.second) {
-      if (Caller >=(CurAddr + Rng.first) &&
-          Caller <= (CurAddr + Rng.second)) {
-        return CurAddr;
-      }
+      if (Caller >= (ImgBase + Rng.first) && Caller <= (ImgBase + Rng.second))
+        return ImgBase;
     }
   }
   return 0;
 }
 
-void RegisterEHFrames(uint8_t* A, size_t Size, uintptr_t BaseAddr, bool Block) {
-  assert(getImageBaseMap().find(DWORD64(A)) == getImageBaseMap().end());
+void RegisterEHFrames(uintptr_t ImgBs, const EHFrameInfos& Frames, bool Block) {
+  if (Frames.empty())
+    return;
+  assert(getImageBaseMap().find(ImgBs) == getImageBaseMap().end());
 
-  PRUNTIME_FUNCTION RFunc = reinterpret_cast<PRUNTIME_FUNCTION>(A);
-  const size_t N = Size / sizeof(RUNTIME_FUNCTION);
-  ImageBaseMap::mapped_type &Ranges = getImageBaseMap()[BaseAddr];
+  ImageBaseMap::mapped_type &Ranges = getImageBaseMap()[ImgBs];
+  ImageRanges::value_type* BlockRange = nullptr;
   if (Block) {
     // Merge all unwind adresses into a single contiguous block
     Ranges.emplace_back(std::numeric_limits<DWORD>::max(),
                         std::numeric_limits<DWORD>::min());
-    ImageRanges::value_type& Range = Ranges.front();
-    for (PRUNTIME_FUNCTION It = RFunc, End = RFunc +N; It < End; ++It) {
-      Range.first = std::min(Range.first, It->BeginAddress);
-      Range.second = std::max(Range.second, It->EndAddress);
+    BlockRange = &Ranges.back();
+  }
+
+  for (auto&& Frame : Frames) {
+    assert(getImageBaseMap().find(DWORD64(Frame.Addr)) ==
+           getImageBaseMap().end() && "Runtime function should not be a key!");
+
+    PRUNTIME_FUNCTION RFunc = reinterpret_cast<PRUNTIME_FUNCTION>(Frame.Addr);
+    const size_t N = Frame.Size / sizeof(RUNTIME_FUNCTION);
+    if (BlockRange) {
+      for (PRUNTIME_FUNCTION It = RFunc, End = RFunc + N; It < End; ++It) {
+        BlockRange->first = std::min(BlockRange->first, It->BeginAddress);
+        BlockRange->second = std::max(BlockRange->second, It->EndAddress);
+      }
+    } else {
+      for (PRUNTIME_FUNCTION It = RFunc, End = RFunc + N; It < End; ++It)
+        Ranges.emplace_back(It->BeginAddress, It->EndAddress);
     }
-  } else {
-    for (PRUNTIME_FUNCTION It = RFunc, End = RFunc +N; It < End; ++It)
-      Ranges.emplace_back(It->BeginAddress, It->EndAddress);
 
-    // Initial sort and merge
-    MergeRanges(Ranges);
+    ::RtlAddFunctionTable(RFunc, N, ImgBs);
   }
 
-  ::RtlAddFunctionTable(RFunc, N, BaseAddr);
+  if (!Block)
+    MergeRanges(Ranges); // Initial sort and merge
 }
 
-void DeRegisterEHFrames(uint8_t* Addr, size_t Size) {
-  assert(getImageBaseMap().find(DWORD64(Addr)) != getImageBaseMap().end());
+void DeRegisterEHFrames(uintptr_t ImgBase, EHFrameInfos Frames) {
```

Comment:
Should be: `void DeRegisterEHFrames(uintptr_t BaseAddr, const EHFrameInfos& Frames) {`